rust/datafusion/src/physical_plan/merge.rs

```diff
@@ -103,33 +106,38 @@ impl ExecutionPlan for MergeExec {
             )),
             1 => {
                 // bypass any threading if there is a single partition
-                self.input.execute(0)
+                self.input.execute(0).await
             }
             _ => {
                 let partitions_per_thread = (input_partitions / self.concurrency).max(1);
                 let range: Vec<usize> = (0..input_partitions).collect();
                 let chunks = range.chunks(partitions_per_thread);
-                let threads: Vec<JoinHandle<Result<Vec<RecordBatch>>>> = chunks
-                    .map(|chunk| {
-                        let chunk = chunk.to_vec();
-                        let input = self.input.clone();
-                        thread::spawn(move || {
-                            let mut batches = vec![];
+
+                let mut tasks = vec![];
+                for chunk in chunks {
+                    let chunk = chunk.to_vec();
+                    let input = self.input.clone();
+                    let task: JoinHandle<Result<Vec<Arc<RecordBatch>>>> =
```

Comment:
Yeah, this is really much nicer in my opinion -- we spawn tasks (not threads) -- and thus we won't create more threads than cpus and give users better control over how the tasks are run. 👍